finality-verifiers/grandpa/src/lib.rs

```diff
@@ -0,0 +1,2046 @@
+    #[pallet::getter(fn get_bridged_block_hash)]
+    pub type BestFinalizedMap<T: Config<I>, I: 'static = ()> =
+        StorageMap<_, Blake2_256, ChainId, BridgedBlockHash<T, I>>;
+
+    /// A ring buffer of imported hashes. Ordered by the insertion time.
+    #[pallet::storage]
+    #[pallet::getter(fn get_multi_imported_hashes)]
+    pub(super) type MultiImportedHashes<T: Config<I>, I: 'static = ()> =
+        StorageDoubleMap<_, Blake2_256, ChainId, Identity, u32, BridgedBlockHash<T, I>>;
+
+    /// Current ring buffer position.
+    #[pallet::storage]
+    #[pallet::getter(fn get_multi_imported_hashes_pointer)]
+    pub(super) type MultiImportedHashesPointer<T: Config<I>, I: 'static = ()> =
+        StorageMap<_, Blake2_256, ChainId, u32>;
+
+    /// Headers which have been imported into the pallet.
+    #[pallet::storage]
+    #[pallet::getter(fn get_multi_imported_headers)]
+    pub type MultiImportedHeaders<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+        _,
+        Blake2_256,
+        ChainId,
+        Identity,
+        BridgedBlockHash<T, I>,
+        BridgedHeader<T, I>,
+    >;
+
+    /// Roots (ExtrinsicsRoot + StateRoot) which have been imported into the pallet for a given gateway.
+    #[pallet::storage]
+    #[pallet::getter(fn get_imported_roots)]
+    pub(super) type MultiImportedRoots<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+        _,
+        Blake2_256,
+        ChainId,
+        Identity,
+        BridgedBlockHash<T, I>,
+        (BridgedBlockHash<T, I>, BridgedBlockHash<T, I>),
+    >;
+
+    #[pallet::storage]
+    pub(super) type RelayChainId<T: Config<I>, I: 'static = ()> =
+        StorageValue<_, ChainId, OptionQuery>;
+
+    /// The current GRANDPA Authority set.
+    #[pallet::storage]
+    pub(super) type CurrentAuthoritySet<T: Config<I>, I: 'static = ()> =
+        StorageValue<_, bp_header_chain::AuthoritySet, OptionQuery>;
+
+    /// The current GRANDPA Authority set.
+    #[pallet::storage]
+    pub(super) type ParachainIdMap<T: Config<I>, I: 'static = ()> =
+        StorageMap<_, Blake2_256, ChainId, Parachain>;
+
+    /// Optional pallet owner.
+    ///
+    /// Pallet owner has a right to halt all pallet operations and then resume it. If it is
+    /// `None`, then there are no direct ways to halt/resume pallet operations, but other
+    /// runtime methods may still be used to do that (i.e. democracy::referendum to update halt
+    /// flag directly or call the `halt_operations`).
+    #[pallet::storage]
+    pub(super) type PalletOwnerMap<T: Config<I>, I: 'static = ()> =
+        StorageMap<_, Blake2_256, ChainId, T::AccountId>;
+
+    /// If true, all pallet transactions are failed immediately.
+    #[pallet::storage]
+    pub(super) type IsHalted<T: Config<I>, I: 'static = ()> = StorageValue<_, bool, ValueQuery>;
+
+    /// If true, all pallet transactions are failed immediately.
+    #[pallet::storage]
+    pub(super) type IsHaltedMap<T: Config<I>, I: 'static = ()> =
```

Comment:
If this is any are true, Is this supposed to be a map?

Reply:
The idea is that every connected chain can be halted, so either the relaychain which in effect stops every parachain, or individual parachains.